### PR TITLE
feat: adding Azure Automation Account module

### LINKED ISSUE #658: Add Azure Automation Account module
This module adds a module for Azure Automation Account.

Examples are divided into sub-tests that include examples that test for the following:

- Automation Account properties such as the Automation Account RunAs account, certificate, and connection
- Module properties in the Automation Account
- DSC properties in the Automation Account
- DSC properties when deployed to a VM and status in Automation Account

### PR DESCRIPTION
**Pull request by robcamer**
Azure Automation Account module for terratest with the following:

(i) terraform example in examples/azure folder
(ii) automationaccount.go module in modules/azure folder
(iii) unit test in test/azure folder
(iv) integration test with example in test/azure folder

NOTE: Running the integration test can take up to 15 minutes.  Use the -timeout flag to set a 15 minute timeout: 
`go test terraform_azure_automationaccount_example_test.go -timeout 15m`

This example requires a service principal and test certificate in order to run successfully.  This is documented in the readme.md for the example:
_"To run this example, you must create a service principal in Azure Active Directory as well as create a non-password protected self-signed certificate in .pfx format that you will need to upload into the Automation Account Run As service principal in Azure Active Directory as a secret for test purposes.

The same certificate file will need to be placed in the `/examples/azure/terraform-azure-automationaccount-example/certificate/` folder with the name `SPRunAsCert.pfx` so that the certificate can be uploaded into the Automation Account in order to successfully configure the Automation Account RunAs account and connection."_

Fixes #658

### CODE REVIEW

**Comment by rguthriemsft** [comment]
@robcamer, I will do an initial review on Monday.  Thank you for your submission.

File: examples/azure/terraform-azure-automationaccount-example/main.tf [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+resource "random_password" "virtual_machine" {
+  length           = 16
+  override_special = "-_%@"
+  min_upper        = "1"
+  min_lower        = "1"
+  min_numeric      = "1"
+  min_special      = "1"
+}
+
+resource "azurerm_automation_account" "automation_account" {
```
Comment (HadwaAbdelhalem) [inline examples/azure/terraform-azure-automationaccount-example/main.tf:42]: consider removing  automation_account_name varibales and set the  name = "terratest-AutomationAccount-${var.postfix}" . like how the rest of the azure examples are implemented.

File: examples/azure/terraform-azure-automationaccount-example/main.tf [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+# # ---------------------------------------------------------------------------------------------------------------------
+# # COMPILE THE SAMPLE DSC IN THE AUTOMATION ACCOUNT
+# # The Terraform `null_resource` signs in to Azure.  The second performs the compliation
+# # Compilation is triggered on every one to ensure the latest changes are always applied
+# # ---------------------------------------------------------------------------------------------------------------------
+
+resource "null_resource" "azureSignInPWSH" {
+  provisioner "local-exec" {
```
Comment (HadwaAbdelhalem) [inline examples/azure/terraform-azure-automationaccount-example/main.tf:65]: ConvertTo-SecureString -String '${var.POWERSHELL_CLIENT_SECRET}, will expose the actual secret value in the logs. consider adding a warning comment

**Review by HadwaAbdelhalem — CHANGES_REQUESTED** [review]

File: examples/azure/terraform-azure-automationaccount-example/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    - `Automation Account Connection Run As Account` with the name specified in the `automation_run_as_connection_name` variable.  See the section titled *_Example Service Principal and Certificate Setup_* below on how to configure the Automation Account RunAs Account credentials.
+    - `Automation Account Connection Run As Certificate` with the name specified in the `automation_run_as_certificate_name` variable and the thumbprint in the `TF_VAR_AUTOMATION_RUN_AS_CERTIFICATE_THUMBPRINT` variable.
+    - `Automation Account Connection Type` with type specified in the `automation_run_as_connection_type` variable.
+    - [Desired State Configuration](https://docs.microsoft.com/en-us/powershell/scripting/dsc/getting-started/winGettingStarted?view=powershell-7#:~:text=Get%20started%20with%20Desired%20State%20Configuration%20%28DSC%29%20for,Windows%20PowerShell%20Desired%20State%20Configuration%20log%20files.%20) with the name specified in the `sample_dsc_name` variable and the path to the DSC specified in the `sample_dsc_path` variable.
+  - [Virtual Machine](https://docs.microsoft.com/en-us/azure/virtual-machines/) with the name specified in the `vm_name` output variable.
```
[the remark this answers is not in the document]
Comment (robcamer) [reply examples/azure/terraform-azure-automationaccount-example/README.md]: Fixed indentation for list to be consistent.

File: examples/azure/terraform-azure-automationaccount-example/main.tf [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+  required_version = ">= 0.12"
+}
+
+# ---------------------------------------------------------------------------------------------------------------------
+# DEPLOY A RESOURCE GROUP
+# See test/terraform_azure_nsg_example_test.go for how to write automated tests for this code.
+# ---------------------------------------------------------------------------------------------------------------------
+
+resource "azurerm_resource_group" "automation_account_dsc_rg" {
+    name     = "${var.resource_group_name}-${var.postfix}"
```
[the remark this answers is not in the document]
Comment (robcamer) [reply examples/azure/terraform-azure-automationaccount-example/main.tf]: removed spaces.  Ran terraform fmt.

File: examples/azure/terraform-azure-automationaccount-example/variables.tf [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+variable "automation_run_as_connection_type" {
+  description = "The name of the automation account run as connection type"
+  type        = string
+  default     = "AzureServicePrincipal"
+}
+
+variable "sample_dsc_name" {
+  description = "The name of the sample DSC configuration that contains the configuraitons that can be applied"
```
[the remark this answers is not in the document]
Comment (robcamer) [reply examples/azure/terraform-azure-automationaccount-example/variables.tf]: Spelling corrected.

File: examples/azure/terraform-azure-automationaccount-example/variables.tf [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+variable "sample_dsc_path" {
+  description = "The path to the sample dsc file in the repo"
+  type        = string
+  default     = "./dsc/SampleDSC.ps1"
+}
+
+variable "sample_dsc_configuration_name" {
+  description = "The name of the DSC configuraiton to apply to the VM"
```
[the remark this answers is not in the document]
Comment (robcamer) [reply examples/azure/terraform-azure-automationaccount-example/variables.tf]: spelling corrected.

File: examples/azure/terraform-azure-automationaccount-example/variables.tf [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  type        = string
+  default     = "SampleDSC.NotWebServer"
+}
+
+variable "vm_name" {
+  description = "The name of the test VM where the DSC configuraiton will be applied"
```
[the remark this answers is not in the document]
Comment (robcamer) [reply examples/azure/terraform-azure-automationaccount-example/variables.tf]: spelling corrected.

File: modules/azure/automationaccount_test.go [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+	require.NoError(t, err)
+	assert.NotEmpty(t, *client)
+}
+
+func TestGetAutomationAccountDscConfigurationE(t, testing.T) {
+	t.Parallel()
+
+	dscConfigurationName = ""
+	automationAccountName := ""
+	resourceGroupName := ""
```
[the remark this answers is not in the document]
Comment (robcamer) [reply modules/azure/automationaccount_test.go:129]: Reviewed all test calls.  Corrected test methods to call tested methods with subscriptionID.  There was a typo in the called method signature, now fixed.

File: test/azure/terraform_azure_automationaccount_example_test.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+func TestTerraformAzureAutomationAccountExample(t *testing.T) {
+	t.Parallel()
+
+	// subscriptionID is overridden by the environment variable "ARM_SUBSCRIPTION_ID"
+	subscriptionID := ""
+	uniquePostfix := random.UniqueId()
+	expectedAutomationAccountName := "terratest-AutomationAccount"
+	expectedSampleDscName := "SampleDSC"
+	expectedSampleDscConfigurationName := "SampleDSC.NotWebServer"
```
[the remark this answers is not in the document]
Comment (robcamer) [reply test/azure/terraform_azure_automationaccount_example_test.go:25]: I ran golint on the /terratest/modules/azure directory.  Only one issue is returned:
recoveryservices.go:37:1: comment on exported function RecoveryServicesVaultExistsE should be of the form "RecoveryServicesVaultExistsE ..."
Go tools are updated.  go version go1.15.6 darwin/amd64

File: test/azure/terraform_azure_automationaccount_example_test.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+	assert.True(t, actualAutomationAccountExists)
+	// Check that the Run As configuraiton is valid
+	runAsAccountValidates := azure.AutomationAccountRunAsConnectionValidates(t, expectedRunAsAccountName+"-"+uniquePostfix, expectedRunAsType, expectedRunAsCertificateThumbprint, automationAccountName, resourceGroupName, subscriptionID)
+	assert.True(t, runAsAccountValidates)
+	//Check that the sample DSC was uploaded successfully into the deployed automation account
+	actualDSCExists := azure.AutomationAccountDscExists(t, expectedSampleDscName, automationAccountName, resourceGroupName, subscriptionID)
+	assert.True(t, actualDSCExists)
+	//Check that the DSC in the automation account successfully compiled
+	dscCompiled := azure.AutomationAccountDscCompiled(t, expectedSampleDscName, automationAccountName, resourceGroupName, subscriptionID)
+	assert.True(t, dscCompiled)
```
[the remark this answers is not in the document]
Comment (robcamer) [reply test/azure/terraform_azure_automationaccount_example_test.go:68]: Removed.

File: examples/azure/terraform-azure-automationaccount-example/main.tf (hunk 1 above)
Comment (robcamer) [reply examples/azure/terraform-azure-automationaccount-example/main.tf:42]: It's a best practice in Terraform to not hard code properties / names for resources and use variables in main.tf.  There are several Azure examples that use variables as well:
terraform-azure-vm-example
terraform-azure-nsg-example
terraform-azure.sqldb-example

I'd prefer to use variables and not hard-code all property values but we if we want to hard-code all property values it should be consistent for all examples.  Should I update the others as well?

File: test/azure/terraform_azure_automationaccount_example_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (HadwaAbdelhalem) [reply test/azure/terraform_azure_automationaccount_example_test.go:25]: @robcamer the test file is not is the modules folder, please run go-lint on the terratest/test/azure/

File: examples/azure/terraform-azure-automationaccount-example/main.tf (hunk 2 above)
Comment (robcamer) [reply examples/azure/terraform-azure-automationaccount-example/main.tf:65]: Added Warning to the header comment for the null_resource.

**Comment by rguthriemsft** [comment]
[Microsoft CI Bot] TL;DR; failure :facepalm:

You can check the status of the CI Pipeline logs here ; https://github.com/rguthriemsft/terratest/actions/runs/665874892

**Comment by robcamer** [comment]
Closing PR due to moving repo to a new location.

**Comment by robcamer** [comment]
Closing this PR as the source repo is moving to a new location.

**Pull request closed without merging**